hudi-utilities/src/test/java/org/apache/hudi/utilities/sources/TestJdbcSource.java

```diff
@@ -0,0 +1,358 @@
+    FileSystem fs = FileSystem.get(new Configuration());
+    FSDataOutputStream outputStream = fs.create(new Path(DatabaseUtils.getFullPathToSecret()));
+    outputStream.writeBytes(DatabaseUtils.getSecret());
+    outputStream.close();
+  }
+
+  public void cleanDerby() {
+    try {
+      Files.walk(Paths.get(DatabaseUtils.getDerbyDir()))
+          .sorted(Comparator.reverseOrder())
+          .map(java.nio.file.Path::toFile)
+          .forEach(File::delete);
+    } catch (Exception e) {
+      //exit silently
+      return;
+    }
+  }
+
+  @Test
+  public void testOnlyInserts() {
+    try {
+      int recordsToInsert = 100;
+      String commitTime = "000";
+
+      //Insert records to Derby
+      DatabaseUtils.clearAndInsert(commitTime, recordsToInsert);
+
+      //Validate if we have specified records in db
+      Assert.assertEquals(recordsToInsert, DatabaseUtils.count());
+
+      //Prepare props
+      clear(props);
+      prepareProps(props);
+      addSecretStringToProps(props);
+
+      //Start JDBCSource
+      Source jdbcSource = new JDBCSource(props, jsc, sparkSession, null);
+      InputBatch inputBatch = jdbcSource.fetchNewData(Option.empty(), Long.MAX_VALUE);
+      Dataset<Row> rowDataset = (Dataset<Row>) inputBatch.getBatch().get();
+
+      //Assert if recordsToInsert are equal to the records in the DF
+      long count = rowDataset.count();
+      LOG.info("Num records in output {}", count);
+      Assert.assertEquals(count, recordsToInsert);
+
+    } catch (IOException | SQLException e) {
+      e.printStackTrace();
+      Assert.fail(e.getMessage());
+    }
+  }
+
+  @Test
+  public void testInsertAndUpdate() throws IOException, SQLException {
+    final String commitTime = "000";
+    final int recordsToInsert = 100;
+    final int recordsToUpdate = 50;
+
+    List<HoodieRecord> insert = DatabaseUtils.clearAndInsert(commitTime, recordsToInsert);
+    List<HoodieRecord> update = DatabaseUtils.update(commitTime, insert);
+  }
+
+  @Test
+  public void testCheckpointingWhenPipelineIsIncremental() throws IOException, SQLException {
+    DatabaseUtils.clearAndInsert("000", 10);
+
+    //Prepare props
+    clear(props);
+    prepareProps(props);
+    addSecretStringToProps(props);
+
+    //Start JDBCSource
+    Source jdbcSource = new JDBCSource(props, jsc, sparkSession, null);
+    InputBatch inputBatch = jdbcSource.fetchNewData(Option.empty(), Long.MAX_VALUE);
+    Dataset<Row> rowDataset = (Dataset<Row>) inputBatch.getBatch().get();
+    String incrementalVal = rowDataset.agg(functions.max(functions.col(
+        props.getString("hoodie.datasource.jdbc.table.incremental.column.name"))).cast(DataTypes.StringType)).first()
+        .getString(0);
+    System.out.println("Incremental value is " + incrementalVal);
+
+    DatabaseUtils.insert("000", 5);
+    Source jdbcSource1 = new JDBCSource(props, jsc, sparkSession, null);
+    InputBatch inputBatch1 = jdbcSource1.fetchNewData(Option.of(incrementalVal), Long.MAX_VALUE);
```

Comment:
The way we do jdbc incremental pull is that we will run use the checkpoint value and prepare a ppd query, the query we prepare inside the JDBCSource with the incremental value is "(select * from table where  incremental_column > '2019-09-09 15:45:01.53' ) rdbms_table" and we except that the incremental column is either a timestamp or an int/long val. 

This worked well when we tried it with MYSQL and mysql was ok converting the string column into int or long or timestamp based on whichever column we compared it with. However, derby is not doing that and is throwing exceptions. if the ppd query is changed to "(select * from table where  incremental_column >Timestamp( '2019-09-09 15:45:01.53') ) rdbms_table" then it works as the incremental column value has been casted/converted to timestamp, but normal strings do not work here.

Reply:
Any help on this will be appreciated!

Reply:
Is it possible to get the type of columns? If it is a timestamp, just convert or cast to Sting?